### ISSUE TITLE
No response from submit turn once game starts

### ISSUE THREAD

**Issue by OrangyTang** [casparluc/VikingDoom#6]
At the moment the server is sending 504 (game not started yet) for a whole bunch of submitted turns, but eventually will just stop responding - I assume something that's supposed to be starting the game seems to be getting stuck.

This is happening with both my own bot and the RandomBot.

**Comment by davinellulinvega** [comment]
Hello,

This was an error on my side in the waiting mechanism I implemented. Your bot and random bot were waiting to play the next game, but the queue raised an exception and so the whole process was forcefully terminated. Which means that no error can be returned and you are simply disconnected.
A solution has been implemented and deployed on the server.

Sincerely.

**Comment by OrangyTang** [comment]
Thanks, I just tried retesting, but I still seem to be having the same issue. With only one bot running it waits for quite a long time, but then eventually a request is sent but no reply is received (it doesn't even time out, which is weird, but that might be my end). Wireshark also shows a request sent (identical  to previous requests) and no response from the server.

What's the actual rules for when I start a game? Will it eventually start a game with one bot, and if so after how long? Or will it always wait for 4 bots before starting one?

**Comment by OrangyTang** [comment]
Also can I suggest that anything that triggers a server side exception should return a 5xx response code, so I know something's gone wrong? Otherwise I can waste a lot of time trying to figure out if somethings just slow (the server seems very laggy sometimes) or has connectivity issues.

Thanks.

**Comment by davinellulinvega** [comment]
Hello,

To answer you question, a game will always wait for all four players to be present before beginning, which might take some time.
Furthermore, at the moment each time an exception is raised in a request or the server is under heavy load, a response with a status **500_internal_server_error** should be sent back to your bot.

Sincerely.

**Comment by OrangyTang** [comment]
Ok, that makes sense. Do you know why the server isn't sending a response once the game starts? It seems to be 100% here.

**Comment by davinellulinvega** [comment]
Hello,

Until all four players are here, the only answer you will get from the server is a response with status **504_gateway_timeout**, no data should be attached. Once all players have joined, you should get a status **200_OK** and the JSON formatted state of the game.
If you look into the code of the RandomBot.py you should see that while the bot is waiting for the game to begin, it simply makes frequent calls to the server, until receiving a status 200.

Sincerely.

**Comment by OrangyTang** [comment]
Sorry I think I wasn't clear in what I'm seeing, both the RandomBot and my own bot are seeing the same thing right now:

1. create game ok
2. POST next turn
3. 504 game not started yet response
4. Repeat 2 and 3 for about 5 minutes
5. POST next turn
6. No response from server

I can only assume that the game is starting on the server, but at that point the 200 response is not actually being sent.

**Comment by davinellulinvega** [comment]
Hello,

This is indeed quite peculiar. I will have to investigate the problem in more depth, because for the moment there is no reason for you to not receive any response from the server. You should at least get a **Timeout**.

Sincerely.

**Comment by davinellulinvega** [comment]
Hello,

To add to the previous comment, I read the server's logs and nothing seems out of the ordinary. Meaning that no processes raised any exception.
Furthermore, the player associated with your user was still in the database and ready to play. Therefore, I am unsure whether the problem comes from the game engine or you own code.

Sincerely.

**Comment by davinellulinvega** [comment]
Hello,

After more investigation on the issue and some testing on my side, I was able to reproduce the error by setting a wait time of 5 seconds between to play request, in the case the server sends a status **504_gateway_timeout** to the first request. The first few requests seem to be working fine a receive a correct response. But once in a while the bot would simply exit.
From what I understand, this could be due to the **keepAlive** configuration of the underlying Apache server. At the moment the **keepAlive** times out after 5 seconds, meaning that with some delay between your computer and the server, at some point the time between two requests will be around 5 seconds and the server will drop the connection (without any error message) and no response will be given to your bot.
To work around this problem, you have the choice between:

- Setting a wait time between request to something less than 5 seconds.
- Implementing a retry mechanism.

Sincerely.

**Comment by davinellulinvega** [comment]
Hello,

For information, the keep-alive configuration has been set to 30 seconds, since removing it completely made the matter worse.

Sincerely.

**Comment by OrangyTang** [comment]
Thanks for the updates. I have added a manual timeout and retry, and with the new server changes things seem a lot more stable.

However I am still unable to actually get into a game. There seem to be a few issues:

1. Old games and players seem to be clogging up games.
I've been launching 4 bots simultaneously, and often two will end up in an 'old' game (that I'd previously created and left) and two will end up in a new game. Thus, none of these bots will ever get out of the 'waiting for game to start' responses.

2. Some kind of race condition with starting new games.
I changed to launching 6 bots simultaneously, some will end up in 'old' games, but the others will end up in completely different games. Thus, none of these games will start either even though there's plenty of bots waiting to start a new game.
If I manually force a delay of 10s between bot launches, then the number of different games seems to reduce, so it definitely looks like a timing issue.

3. Full games not starting
With 6 bots, a staggered start and lots of retries, eventually I got 4 bots in the same game (ie. same game code and map). However all bots continued to receive 'game not started' responses for +5 minutes until I switched them off.

4. No diagnostic for game start
At the moment the 504/not started response doesn't tell me what's actually causing the game to not start. A response message that indicates the problem would be helpful (eg. 3 players joined, waiting for 1 more player... / game starting in 30s / etc.).

Thanks.

**Comment by davinellulinvega** [comment]
Hello,

I am glad to hear that you are able to communicate with the server without further problems.
Concerning your other complaints:

1. A Garbage collection mechanism has been implemented to remove empty games, but since I have not had time to fully test it for the moment it is not deployed on the server.

2. Viking Doom is based on [Thespian](https://godaddy.github.io/Thespian/doc/), which implements the [Actor model](https://en.wikipedia.org/wiki/Actor_model). According to this model, actors are independent computing units that can communicate with other actors through message only. However, an Actor will never wait for an answer to a message. Thus, if you launch 6 bots simultaneously, they will all ask for a game, at which point the system will process all six messages, sometimes creating a new game, sometimes adding to an incomplete game, depending on how fast the Actor responsible for initialising games is able to completely process request for new games. I would not call it a race condition, since no information is being overwritten, it is simply a result of the multi-process architecture. If you wish for you 4 or 6 bots to be in the same game, launch them with a 1-2 seconds delay in between.

3. For the moment, the server on which Viking Doom is running is not all powerful and is alone to manage everything from running the game, to serving the website and running the database. This is all quite demanding in CPU power and for this reason, the number of games that can be played in parallel is limited to 2. If more games are created in the meantime, they will be put in a queue and launched once one of the running games is complete. Thus, you might have to wait up to 15 mins, if the two games just began. If you want to be able to test your strategy in a more flexible environment, you are free to download and install the Viking Doom project on your computer.

4. It is not possible for the system to return a more precise message. Since Viking Doom uses the Actor model and actors do not know of other actors' internal state, the actor in charge of keeping a table of all running games with their corresponding addresses has no way of knowing why a game is not running. In fact since responsibilities are spread across multiple actors, there is no single actor able to give you that kind of precise answer without sending messages to others.

Since you seem to be able to connect without further problems, I will consider this issue as being solved and close it. If you find other errors, please feel free to open a new issue.

Sincerely.
